pkg/agent/controller/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,879 @@
+				if c.numCaptures < defaultWorkers && c.waitingCaptures == 0 {
+					c.numCaptures += 1
+					// run capture asynchronously
+					c.startPacketCapture(pcName)
+					return "RUNNING", nil
+				}
+				// non blocking channel write
+				select {
+				case c.waitingCapturesCh <- pcName:
+					c.waitingCaptures += 1
+					return "WAITING", nil
+				default:
+					// should never happen with realistic usage, will be requeued
+					return "", fmt.Errorf("too many captures in waiting channel")
+				}
+			}()
+			if err != nil {
+				return "", err
+			}
+			c.captures[pcName] = newStatus
+			return newStatus, nil
+		}
+		return status, nil
+	}()
+	if err != nil {
+		return err
+	}
+	if err := c.patchStatusIfNeeded(pcName, status); err != nil {
+		return fmt.Errorf("error when patching status: %w", err)
+	}
+	return nil
+}
+
+func (c *Controller) patchStatusIfNeeded(name string, status string) error {
+	klog.InfoS("Patching PacketCapture CR Status if needed", "name", name, "status", status)
+	conditions := []crdv1alpha1.PacketCaptureCondition{}
+
+	if status == "RUNNING" {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCaptureRunning,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else if status == "WAITING" || status == "" {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCapturePending,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else {
+		return nil
+	}
+
+	toUpdate, getErr := c.packetCaptureLister.Get(name)
+	if getErr != nil {
+		return getErr
+	}
+	if retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		newCond := mergeConditions(toUpdate.Status.Conditions, conditions)
+		if conditionSliceEqualsIgnoreLastTransitionTime(toUpdate.Status.Conditions, newCond) {
+			return nil
+		}
+		toUpdate.Status.Conditions = newCond
+		klog.V(2).InfoS("Updating PacketCapture", "name", name, "status", toUpdate.Status)
+		_, updateErr := c.crdClient.CrdV1alpha1().PacketCaptures().UpdateStatus(context.TODO(), toUpdate, metav1.UpdateOptions{})
+		if updateErr != nil && apierrors.IsConflict(updateErr) {
+			var getErr error
+			if toUpdate, getErr = c.crdClient.CrdV1alpha1().PacketCaptures().Get(context.TODO(), name, metav1.GetOptions{}); getErr != nil {
+				return getErr
+			}
+		}
+		return updateErr
+	}); retryErr != nil {
+		return retryErr
+	}
+
+	return nil
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (ftp.Uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func (c *Controller) generatePacketsPathForServer(name string) string {
+	return name + ".pcapng"
+}
+
+func getDefaultFileServerAuth() *crdv1alpha1.BundleServerAuthConfiguration {
```

Comment:
the use of a support-bundle specific API type here doesn't make sense IMO

Reply:
i think i treat `Bundle` (instead of `SupportBundle`) as a generic name in this case, so it doesn't seems too weird for me. sames to `ParseBundleAuth`.

To refactor this, i think a generic struct in the `ftp` package should be fine.

Reply:
updated with a more generic name.